### PR TITLE
Bluetooth ipc

### PR DESCRIPTION
**Pull request by dati91**

### CODE REVIEW

File: components/script/dom/bluetooth.rs [hunk 1]
```diff
@@ -33,7 +38,55 @@ impl BluetoothMethods for Bluetooth {
...
+                                                    let ttx_power = tx_power;
+                                                    let trssi = rssi;
+                                                    let ad_data = &BluetoothAdvertisingData::new(self.global().r(),
+                                                                                                tappearance,
+                                                                                                ttx_power,
+                                                                                                trssi);
+                                                    let vendor_id_source = match tvendor_id_source.as_ref() {
+                                                         "bluetooth" => VendorIDSource::Bluetooth,
+                                                         "usb" => VendorIDSource::Usb,
+                                                         // TODO add unknown choice
```
Comment (dati91) [inline components/script/dom/bluetooth.rs]: remove this line

File: components/script/dom/window.rs [hunk 2]
```diff
@@ -205,10 +206,15 @@ pub struct Window {
     /// The current size of the window, in pixels.
     window_size: Cell<Option<WindowSizeData>>,
 
+
```
Comment (dati91) [inline components/script/dom/window.rs]: remove this line

File: components/net/bluetooth_thread.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+use std::string::String;
+use util::thread::spawn_named;
+
+pub trait BluetoothThreadFactory {
+    fn new() -> Self;
+}
+
+impl BluetoothThreadFactory for BluetoothThread {
+    fn new() -> BluetoothThread {
+        let (sender, reciever) = ipc::channel().unwrap();
```
Comment (zakorgy) [inline components/net/bluetooth_thread.rs]: azt nem receiver-nek kellene írni?

File: components/script/dom/bluetoothdevice.rs [hunk 4]
```diff
@@ -47,7 +46,7 @@ impl BluetoothDevice {
             vendorID: vendorID,
             productID: productID,
             productVersion: productVersion,
-            gatt: MutHeap::new(gatt),
+            gatt: MutNullableHeap::default(),
```
Comment (zakorgy) [inline components/script/dom/bluetoothdevice.rs]: lehetne gatt: Default::default(), is

File: components/script/dom/bluetoothremotegattservice.rs [hunk 5]
```diff
@@ -64,7 +67,50 @@ impl BluetoothRemoteGATTServiceMethods for BluetoothRemoteGATTService {
...
+                                                                                 tindicate,
+                                                                                 tauthenticated_signed_writes,
+                                                                                 treliable_write,
+                                                                                 twritable_auxiliaries);
+                        Some(BluetoothRemoteGATTCharacteristic::new(self.global().r(),
+                                                                    &self,
+                                                                    DOMString::from(tchar_uuid),
+                                                                    properties))
+                    },
+            BluetoothThings::Error { error } => { println!("{}", error);
```
Comment (zakorgy) [inline components/script/dom/bluetoothremotegattservice.rs]: => után mehetne új sorba, mint ahogy bluetoothremotegattserver.rs-ben is van

File: components/net_traits/bluetooth_thread.rs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    GetPrimaryService(IpcSender<BluetoothThings>),
+    GetDescriptor(IpcSender<BluetoothThings>),
+    GATTServerConnect(IpcSender<BluetoothThings>),
+    ReadValue(IpcSender<BluetoothThings>),
+    WriteValue(IpcSender<BluetoothThings>),
+    Exit,
+}
+
+#[derive(Deserialize, Serialize)]
+pub enum BluetoothThings {
```
Comment (dati91) [inline components/net_traits/bluetooth_thread.rs]: We should rename this!
Something like: `BluetoothMsg` or simply `Bluetooth`

File: components/net_traits/bluetooth_thread.rs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        /* Bluetooth Device properties */
+        id: String, name: String, device_class: u32,
+        vendor_id_source: String, vendor_id: u32,
+        product_id: u32, product_version: u32,
+        /* Advertisiong Data properties */
+        appearance: u16, tx_power: i8, rssi: i8 },
+     BluetoothCharacteristic {
+        /*Characteristic properties*/
+        char_uuid: String,
+        /*CharacteristicProperties*/
```
Comment (dati91) [inline components/net_traits/bluetooth_thread.rs]: remove this line

File: components/script/dom/bluetoothremotegattcharacteristic.rs [hunk 8]
```diff
@@ -70,8 +72,24 @@ impl BluetoothRemoteGATTCharacteristicMethods for BluetoothRemoteGATTCharacteris
...
     // https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothremotegattcharacteristic-getdescriptor
     fn GetDescriptor(&self) -> Option<Root<BluetoothRemoteGATTDescriptor>> {
-        //UNIMPLEMENTED
-        None
+        let (sender, receiver) = ipc::channel().unwrap();
+        self.global().r().as_window().bluetooth_thread().send(BluetoothThreadMsg::GetDescriptor(sender)).unwrap();
+        let btd = receiver.recv().unwrap();
+        match btd {
+            BluetoothThings::BluetoothDescriptor { uuid } =>
+                {
```
Comment (zakorgy) [inline components/script/dom/bluetoothremotegattcharacteristic.rs]: unintend this line with one Tab?

File: components/script/dom/bluetoothremotegattcharacteristic.rs [hunk 9]
```diff
@@ -70,8 +72,24 @@ impl BluetoothRemoteGATTCharacteristicMethods for BluetoothRemoteGATTCharacteris
...
+                    let tuuid = uuid;
+                    Some(BluetoothRemoteGATTDescriptor::new(self.global().r(),
+                                                            &self,
+                                                            DOMString::from(tuuid)))
+                },
```
Comment (zakorgy) [inline components/script/dom/bluetoothremotegattcharacteristic.rs]: same

File: components/script/dom/bluetoothremotegattcharacteristic.rs [hunk 10]
```diff
@@ -70,8 +72,24 @@ impl BluetoothRemoteGATTCharacteristicMethods for BluetoothRemoteGATTCharacteris
...
+            BluetoothThings::Error { error } =>
+                {
```
Comment (zakorgy) [inline components/script/dom/bluetoothremotegattcharacteristic.rs]: same

File: components/script/dom/bluetoothremotegattcharacteristic.rs [hunk 11]
```diff
@@ -70,8 +72,24 @@ impl BluetoothRemoteGATTCharacteristicMethods for BluetoothRemoteGATTCharacteris
...
+                    println!("{}", error);
+                    None
+                },
```
Comment (zakorgy) [inline components/script/dom/bluetoothremotegattcharacteristic.rs]: same

File: components/net/bluetooth_thread.rs [hunk 12]
```diff
@@ -0,0 +1,128 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use ipc_channel::ipc::{self, IpcReceiver, IpcSender};
+use net_traits::bluetooth_thread::{BluetoothThreadMsg, BluetoothThread, BluetoothMsg};
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: use net_traits::bluetooth_thread::{BluetoothThread, BluetoothThreadMsg, BluetoothMsg};

File: components/net/bluetooth_thread.rs [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+                BluetoothThreadMsg::Exit => {
+                    break
+                }
+            }
+        }
+    }
+
+    fn request_device(&self, sender: IpcSender<BluetoothMsg>) {
+        let message = BluetoothMsg::BluetoothDevice {
+            id: String::from("ca:fe:c0:ff:ee:00"),
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs:69]: From where do these values come from? Maybe add some comments.

File: components/net/bluetooth_thread.rs [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+            appearance: 832,
+            tx_power: 40,
+            rssi: -65
+        };
+        sender.send(message).unwrap();
+    }
+
+    pub fn get_characteristic(&self, sender: IpcSender<BluetoothMsg>) {
+        let message = BluetoothMsg::BluetoothCharacteristic {
+            uuid: String::from("00002a19-0000-1000-8000-00805f9b34fb"),
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs:97]: Same

File: components/net/bluetooth_thread.rs [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+            authenticated_signed_writes: false,
+            reliable_write: false,
+            writable_auxiliaries: false
+        };
+        sender.send(message).unwrap();
+    }
+
+    pub fn get_primary_service(&self, sender: IpcSender<BluetoothMsg>) {
+        let message = BluetoothMsg::BluetoothService {
+            uuid: String::from("0000180f-0000-1000-8000-00805f9b34fb"),
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: Same

File: components/net/bluetooth_thread.rs [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+            is_primary: true
+        };
+        sender.send(message).unwrap();
+    }
+
+    pub fn get_descriptor(&self, sender: IpcSender<BluetoothMsg>) {
+        let message = BluetoothMsg::BluetoothDescriptor {
+            uuid: String::from("00002902-0000-1000-8000-00805f9b34fb")
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: Same

File: components/net/bluetooth_thread.rs [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+        };
+        sender.send(message).unwrap();
+    }
+
+    pub fn gatt_server_connect(&self, _sender: IpcSender<BluetoothMsg>) {
+        //UNIMPLEMENTED
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: Is this comment necessary?

File: components/net/bluetooth_thread.rs (hunk 16 above)
Comment (mmatyas) [inline components/net/bluetooth_thread.rs:113]: "00002902-0000-1000-8000-00805f9b34fb" appears often, consider storing it as a static string.

File: components/net_traits/bluetooth_thread.rs [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+pub enum BluetoothThreadMsg {
+    RequestDevice(IpcSender<BluetoothMsg>),
+    GetCharacteristic(IpcSender<BluetoothMsg>),
+    GetPrimaryService(IpcSender<BluetoothMsg>),
+    GetDescriptor(IpcSender<BluetoothMsg>),
+    GATTServerConnect(IpcSender<BluetoothMsg>),
+    ReadValue(IpcSender<BluetoothMsg>),
+    WriteValue(IpcSender<BluetoothMsg>),
+    Exit,
+}
```
Comment (mmatyas) [inline components/net_traits/bluetooth_thread.rs:16]: Are these sorted in the order you supposed to use the messages?

File: components/net_traits/bluetooth_thread.rs [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+
+#[derive(Deserialize, Serialize)]
+pub enum BluetoothMsg {
+    BluetoothDevice {
+        /* Bluetooth Device properties */
```
Comment (mmatyas) [inline components/net_traits/bluetooth_thread.rs]: `//` is preferred for comments in Rust code

File: components/script/dom/bluetooth.rs [hunk 20]
```diff
@@ -4,10 +4,15 @@
 
 use dom::bindings::codegen::Bindings::BluetoothBinding;
 use dom::bindings::codegen::Bindings::BluetoothBinding::BluetoothMethods;
+use dom::bindings::codegen::Bindings::BluetoothDeviceBinding::VendorIDSource;
 use dom::bindings::global::GlobalRef;
 use dom::bindings::js::Root;
-use dom::bindings::reflector::{Reflector, reflect_dom_object};
+use dom::bindings::reflector::{Reflector, Reflectable, reflect_dom_object};
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: Maybe put Reflectable first

File: components/script/dom/bluetooth.rs [hunk 21]
```diff
@@ -33,7 +38,50 @@ impl BluetoothMethods for Bluetooth {
 
     // https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetooth-requestdevice
     fn RequestDevice(&self) -> Option<Root<BluetoothDevice>> {
-        //UNIMPLEMENTED
-        None
+        let (sender, receiver) = ipc::channel().unwrap();
+        self.global().r().as_window().bluetooth_thread().send(BluetoothThreadMsg::RequestDevice(sender)).unwrap();
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: I guess this code is still WIP?

File: components/script/dom/bluetooth.rs [hunk 22]
```diff
@@ -33,7 +38,50 @@ impl BluetoothMethods for Bluetooth {
...
+        let btd = receiver.recv().unwrap();
+        //(id, name, device_class, vendor_id_source, vendor_id, product_id, product_version
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: Unnecessary.

File: components/script/dom/bluetooth.rs [hunk 23]
```diff
@@ -33,7 +38,50 @@ impl BluetoothMethods for Bluetooth {
...
+                                          device_class,
+                                          vendor_id_source,
+                                          vendor_id,
+                                          product_id,
+                                          product_version))
+            },
+            BluetoothMsg::Error {
+                error
+            } => {
+                println!("{}", error);
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs:79]: There's a better way to handle errors.

File: components/script/dom/bluetoothremotegattcharacteristic.rs [hunk 24]
```diff
@@ -4,14 +4,16 @@
 
 use dom::bindings::codegen::Bindings::BluetoothRemoteGATTCharacteristicBinding;
 use dom::bindings::codegen::Bindings::BluetoothRemoteGATTCharacteristicBinding::
-    BluetoothRemoteGATTCharacteristicMethods;
+BluetoothRemoteGATTCharacteristicMethods;
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattcharacteristic.rs]: I'd prefer the indent, why was it removed? tidy?

File: components/script/dom/bluetoothremotegattcharacteristic.rs [hunk 25]
```diff
@@ -70,8 +72,25 @@ impl BluetoothRemoteGATTCharacteristicMethods for BluetoothRemoteGATTCharacteris
 
     // https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothremotegattcharacteristic-getdescriptor
     fn GetDescriptor(&self) -> Option<Root<BluetoothRemoteGATTDescriptor>> {
-        //UNIMPLEMENTED
-        None
+        let (sender, receiver) = ipc::channel().unwrap();
+        self.global().r().as_window().bluetooth_thread().send(BluetoothThreadMsg::GetDescriptor(sender)).unwrap();
+        let btd = receiver.recv().unwrap();
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattcharacteristic.rs]: Nice unwraps.

File: components/script/dom/bluetoothremotegattcharacteristic.rs [hunk 26]
```diff
@@ -70,8 +72,25 @@ impl BluetoothRemoteGATTCharacteristicMethods for BluetoothRemoteGATTCharacteris
...
+                uuid
+            } => {
+                Some(BluetoothRemoteGATTDescriptor::new(self.global().r(),
+                                                        &self,
+                                                        DOMString::from(uuid)))
+            },
+            BluetoothMsg::Error {
+                error
+            } => {
+                println!("{}", error);
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattcharacteristic.rs:89]: Again, check error handling in other components.

File: components/script/dom/bluetoothremotegattserver.rs [hunk 27]
```diff
@@ -20,18 +23,17 @@ pub struct BluetoothRemoteGATTServer {
...
+            connected: Cell::new(false),
         }
     }
 
-    pub fn new(global: GlobalRef, device: &BluetoothDevice, connected: bool) -> Root<BluetoothRemoteGATTServer> {
+    pub fn new(global: GlobalRef, device: &BluetoothDevice) -> Root<BluetoothRemoteGATTServer> {
         reflect_dom_object(box BluetoothRemoteGATTServer::new_inherited(
-                           device,
-                           connected),
+                           device),
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattserver.rs]: put this in the same line if this is the only value

File: components/script/dom/bluetoothremotegattserver.rs [hunk 28]
```diff
@@ -64,7 +66,26 @@ impl BluetoothRemoteGATTServerMethods for BluetoothRemoteGATTServer {
 
     // https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothremotegattserver-getprimaryservice
     fn GetPrimaryService(&self) -> Option<Root<BluetoothRemoteGATTService>> {
-        //UNIMPLEMENTED
-        None
+       let (sender, receiver) = ipc::channel().unwrap();
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattserver.rs]: Incorrect indentation, I'm surprised tidy didn't pick it up.

File: components/script/dom/bluetoothremotegattserver.rs [hunk 29]
```diff
@@ -64,7 +66,26 @@ impl BluetoothRemoteGATTServerMethods for BluetoothRemoteGATTServer {
...
+        self.global().r().as_window().bluetooth_thread().send(BluetoothThreadMsg::GetPrimaryService(sender)).unwrap();
+        let bts = receiver.recv().unwrap();
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattserver.rs]: :frowning:

File: components/script/dom/bluetoothremotegattserver.rs [hunk 30]
```diff
@@ -64,7 +66,26 @@ impl BluetoothRemoteGATTServerMethods for BluetoothRemoteGATTServer {
...
+            } => {
+                Some(BluetoothRemoteGATTService::new(self.global().r(),
+                                                     &self.device.get(),
+                                                     DOMString::from(uuid),
+                                                     is_primary))
+            },
+            BluetoothMsg::Error {
+                error
+            } => {
+                println!("{}", error);
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattserver.rs:84]: Error handling.

File: components/script/dom/bluetoothremotegattservice.rs [hunk 31]
```diff
@@ -64,7 +67,44 @@ impl BluetoothRemoteGATTServiceMethods for BluetoothRemoteGATTService {
 
     // https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothremotegattservice-getcharacteristic
     fn GetCharacteristic(&self) -> Option<Root<BluetoothRemoteGATTCharacteristic>> {
-        // UNIMPLEMENTED
-        None
+        let (sender, receiver) = ipc::channel().unwrap();
+        self.global().r().as_window().bluetooth_thread().send(BluetoothThreadMsg::GetCharacteristic(sender)).unwrap();
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattservice.rs]: See above

File: components/script/dom/bluetoothremotegattservice.rs [hunk 32]
```diff
@@ -64,7 +67,44 @@ impl BluetoothRemoteGATTServiceMethods for BluetoothRemoteGATTService {
...
+        let btc = receiver.recv().unwrap();
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattservice.rs]: Isn't there a better name than btc?

File: components/script/dom/bluetoothremotegattservice.rs [hunk 33]
```diff
@@ -64,7 +67,44 @@ impl BluetoothRemoteGATTServiceMethods for BluetoothRemoteGATTService {
...
+                                                                         writable_auxiliaries);
+                Some(BluetoothRemoteGATTCharacteristic::new(self.global().r(),
+                                                            &self,
+                                                            DOMString::from(uuid),
+                                                            properties))
+            },
+            BluetoothMsg::Error {
+                error
+            } => {
+                println!("{}", error);
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattservice.rs:104]: Handling errors

File: components/script/dom/webidls/BluetoothDevice.webidl [hunk 34]
```diff
@@ -8,7 +8,8 @@
 // Allocation authorities for Vendor IDs:
 enum VendorIDSource {
     "bluetooth",
-    "usb"
+    "usb",
+    "unknown"
```
Comment (mmatyas) [inline components/script/dom/webidls/BluetoothDevice.webidl:12]: What's unknown?

Comment (dati91) [inline components/script/dom/webidls/BluetoothDevice.webidl:12]: Unknown the source, when we can't read that property

**Pull request merged**